torch/csrc/autograd/function.h

```diff
@@ -151,24 +151,21 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     // probably operate with names.
     at::NoNamesGuard no_names_guard;
 
-    bool pre_sampled = false;
-    if (at::shouldRunRecordFunction(&pre_sampled)) {
-      // Using RecordFunction to trigger observers in the backward pass
-      at::RecordFunction guard(at::RecordScope::BACKWARD_FUNCTION, pre_sampled);
-      if (guard.isActive()) {
-        // Using sequence number and thread id to correlate with
-        // the forward pass function
-        guard.setForwardThreadId(thread_id_);
-        if (guard.needsInputs()) {
-          guard.before(
-            name(),
-            std::vector<c10::IValue>(inputs.begin(), inputs.end()),
-            sequence_nr());
-        } else {
-          guard.before(name(), sequence_nr());
-        }
+    auto step_callbacks = at::getStepCallbacks(at::RecordScope::BACKWARD_FUNCTION);
```

Comment:
As discussed in another duplicate PR: How expensive is this? This is very hot path

Reply:
It is perf neutral when profiling is disabled, and a win when profiling is enabled. (There are benchmarks for the dispatcher which is using the identical pattern.) There's already a good description for why the latter is faster, but let me briefly discuss the former.

#### Not profiling (The old way):
* One TLS lookup (rf_tls)
* Three predictably false branches in `shouldRunRecordFunction`
* An atomic read (global_record_all_functions_)
* A second TLS lookup (coinflip_tls)
* Counter decrement (coinflip_tls)

#### Not profiling (The new way)
* One atomic read (global version counter)
* One TLS lookup (TLS manager)
* One decrement on the TLS
* Copy step_callbacks

So you can see that we're trading more branching and TLS with materializing the active callbacks. This is fine because the active callbacks is just a pair of pointers, so `step_callbacks` is effectively a `SmallVector<pair<void*, void*>, 4`. (plus a couple bools) And so it evens out.

By the way, we could make this implementation *even* faster. In the old formulation the soft limit (4) on callbacks was for *registered* callbacks. So if you registered five callbacks you were going to have a bad time. Here it's for callbacks that fire (plus we can now unregister global callbacks), so the limit is for callbacks that actually fire on a given run, and the sampling callbacks almost never fire. So in practice you should never actually have more than four callbacks that run on a particular invocation. If we make it a *hard* limit we can switch to `std::array<pair<void*, void*>, kHardLimit>` and completely skip `SmallVector::Operator()=` which is the most expensive part of this. I didn't do it here, but something to consider.

Reply:
Thanks for the detailed explanation!